Merge and improve torch optim optimizer type stubs

ringohoffman opened the pull request:
Fixes #102428

Also improves hook registration type hints:

```python
from typing import Any, Dict, Tuple

from torch import nn
from torch.optim import Adam, Adagrad, Optimizer

linear = nn.Linear(2,2)
optimizer = Adam(linear.parameters(), lr=0.001)


def pre_hook_fn_return_none(optimizer: Adam, inputs: Tuple[Any, ...], kwargs: Dict[str, Any]) -> None:
    return None


def pre_hook_fn_return_modified(
    optimizer: Optimizer, inputs: Tuple[Any, ...], kwargs: Dict[str, Any]
) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
    return inputs, kwargs


def hook_fn(optimizer: Optimizer, inputs: Tuple[Any, ...], kwargs: Dict[str, Any]) -> None:
    return None


def hook_fn_other_optimizer(optimizer: Adagrad, inputs: Tuple[Any, ...], kwargs: Dict[str, Any]) -> None:
    return None


optimizer.register_step_post_hook(hook_fn)  # OK

optimizer.register_step_pre_hook(pre_hook_fn_return_none)  # OK
optimizer.register_step_pre_hook(pre_hook_fn_return_modified)  # OK

optimizer.register_step_post_hook(hook_fn_other_optimizer)  # Parameter 1: type "Adam" cannot be assigned to type "Adagrad"

```

Commit:
Merge torch.optim.optimizer type stubs

Fixes #102428

Commit:
Add type: ignore for solved type error regarding Self type

This error has been fixed in newer versions of mypy but is still present in 0.960

Commit:
Rename OptimizerHook to OptimizerPostHook

Commit:
Fix exposed type errors from torch.optim.optimizer type improvements

Commit:
Improve ZeroRedundancyOptimizer.step type hint

use overload, move override ignore since we have an extra kwarg compared to the parent signature

Commit:
Remove overload implementation from stub file

Commit:
Make ParamSpec private to appease test_public_bindings

ringohoffman commented:
```
2023-05-31T05:06:48.4699555Z AssertionError: False is not true : All the APIs below do not meet our guidelines for public API from https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation.
2023-05-31T05:06:48.4700262Z Make sure that everything that is public is expected (in particular that the module has a properly populated `__all__` attribute) and that everything that is supposed to be public does look public (it does not start with `_` and has a `__module__` that is properly populated).
2023-05-31T05:06:48.4700715Z 
2023-05-31T05:06:48.4700784Z Full list:
2023-05-31T05:06:48.4700988Z # torch.optim.optimizer.P:
2023-05-31T05:06:48.4701344Z   - Is NOT public: it is not inside the module's (`torch.optim.optimizer`) `__all__`
2023-05-31T05:06:48.4701859Z   - Does look public: it does look public because it follows the rules from the doc above (does not start with `_` and has a proper `__module__`).
2023-05-31T05:06:48.4702292Z   - You can do either of these two things to fix this problem:
2023-05-31T05:06:48.4702668Z     - To make it public: add it from the modules's (`torch.optim.optimizer`) `__all__`
2023-05-31T05:06:48.4703039Z     - To make it NOT look public: make its name start with `_`
```

This error seems strange since all the TypeVars follow the same format. Does this have something to do with being imported from `typing_extensions`?

albanD reviewed (commented):
Will let @janeyx99 handle this one!

ringohoffman commented:
> > This error seems strange since all the TypeVars follow the same format. Does this have something to do with being imported from `typing_extensions`?
> 
> It has to do with `__all__` not being defined so it Python defaults to the behavior of exporting everything that doesn't begin with a `_`. Our linter indicates that this is likely an error (which it is) so the solution is either to define an `__all__` or rename the typevar as `_P`

It is defined: https://github.com/pytorch/pytorch/pull/102593/files#diff-8015a6c74af00fb9fd8c6fc7f7e5b206f8f8cce0ea6fe2f0653bf62c146c3892R24. Not sure why it isn't an error for all the TypeVars since none of them start with `_`.

Commit:
Make _ConstructableIterable a runtime_checkable Protocol

Creating subclasses of collections.abc classes is actually pretty limited in actual usefulness since creating a subclass of e.g. Iterable doesn't mean that a list will be also a subclass of that subclass; runtime_checkable Protocols give us the intended behavior

Commit:
Use invariant type for Generic Protocol

ringohoffman commented:
@janeyx99 how is this looking? any concerns?

Commit:
Merge branch 'main' into merge-and-improve-torch-optim-optimizer-type-stubs

Commit:
Improve type hints for _group_tensors_by_device_and_dtype

this wasn't fun I'm ngl

ringohoffman commented:
@janeyx99 do you know what time frame this might get prioritized in?

janeyx99 commented:
@ringohoffman hey! sorry about the delay and thanks for the reminder.

janeyx99 commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -144,14 +144,22 @@ def state_dict(self) -> Dict[str, Any]:
+    def step(self, closure: None = ...) -> None:
+        ...
+
+    @overload
```
Could you explain why these two overloads aren't already included in the actual step signature below?

ringohoffman replied:
Only overload signatures are used for signature matching. The last signature without the `@overload` is purely for type hinting.

ringohoffman replied:
The 2 different signatures say this:

1. The `ellipsis` means that the default argument should be matched i.e. `.step()`. Overall this signature says that if the closure is `None`, i.e. `.step()`, `.step(closure=None)`, or `.step(None)`, then we know that `None` will be returned.
```python
@overload
    def step(self, closure: None = ...) -> None:
        ...
```
2. If a `Callable` that accepts no positional or keyword arguments and returns a `float` is provided as the closure, i.e. `.step(closure=lambda: 0.0)` or `.step(lambda: 0.0)`, then we know a `float` will be returned.
```python
    @overload
    def step(self, closure: Callable[[], float]) -> float:
        ...
```

janeyx99 replied:
Thanks so much for the exhaustive explanation!

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -152,6 +225,10 @@ def register_optimizer_step_post_hook(hook: Callable[..., None]) -> RemovableHan
     return handle
 
 
+_P = ParamSpec("_P")
+R = TypeVar("R")
```
Could we move these lower + closer to where they are used?

ringohoffman replied:
I don't think I can move these any closer, they are used in the `staticmethod` `profile_hook_step` and since they are referenced within the method itself, I don't think I can make these type variables a class variables of `Optimizer`.

Did you have something else in mind?

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -265,22 +350,24 @@ def _optimizer_step_code(self):
-            self, *_ = args
+        def wrapper(*args: _P.args, **kwargs: _P.kwargs) -> R:
+            self = args[0]
+            assert isinstance(self, Optimizer)
```
Again, not sure if adding asserts will mess with BC--is there a reason to add this?

ringohoffman replied:
`self` is inferred as `object`

```console
$ mypy torch/optim/optimizer.py
torch/optim/optimizer.py:320:77: error: "object" has no attribute "_optimizer_step_pre_hooks"  [attr-defined]
torch/optim/optimizer.py:331:17: error: "object" has no attribute "_optimizer_step_code"  [attr-defined]
torch/optim/optimizer.py:334:40: error: "object" has no attribute "_optimizer_step_post_hooks"  [attr-defined]
```

These errors indicate to me that this code would certainly fail if it were not an `Optimizer`.

The only apparent usage in all of pytorch: https://github.com/search?q=repo%3Apytorch%2Fpytorch%20profile_hook_step&type=code

is in wrapping the `step` method of this class:

https://github.com/pytorch/pytorch/blob/cf8af57c4a53ea1f8947d90c26340d1b81ca0e19/torch/optim/optimizer.py#L317

where `args[0]` is necessarily an `Optimizer`.

janeyx99 replied:
Alright, I do still believe there is slight BC risk here due to python ducktyping (e.g., a user may have an Optimizer_ class that somehow implements all these methods but does not subclass from Optimizer itself) but this may be a reasonably tiny risk.

ringohoffman replied:
I will use `cast` if there is a concern for BC.

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -294,25 +381,33 @@ def wrapper(*args, **kwargs):
         """Groups a list of lists of tensors by device and dtype.
         Skips this step if we are compiling since this will occur during inductor lowering."""
         if is_compiling():
-            if with_indices:
```
I'd prefer the old if statement instead of the ternary op because in the old implementation, indices doesn't get materialized if with_indices is false. There is a perf implication here.

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -294,25 +381,33 @@ def wrapper(*args, **kwargs):
-    def register_step_pre_hook(self, hook: Callable[..., None]) -> RemovableHandle:
+    def register_step_pre_hook(
+        self,
+        hook: Callable[[Self, Tuple[Any, ...], Dict[str, Any]], Optional[Tuple[Any, Dict[str, Any]]]],  # type: ignore[valid-type]
```
Why not use OptimizerPrehook?

ringohoffman replied:
addressed

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -326,15 +421,18 @@ def register_step_pre_hook(self, hook: Callable[..., None]) -> RemovableHandle:
-    def register_step_post_hook(self, hook: Callable[..., None]) -> RemovableHandle:
+    def register_step_post_hook(
+        self,
+        hook: Callable[[Self, Tuple[Any, ...], Dict[str, Any]], None],  # type: ignore[valid-type]
```
Why not use the type alias OptimizerPostHook?

ringohoffman replied:
addressed with class variable `TypeAlias`

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -467,8 +547,11 @@ def zero_grad(self, set_to_none: bool = True):
-        if foreach:
-            per_device_and_dtype_grads = defaultdict(lambda: defaultdict(list))
+
+        per_device_and_dtype_grads: DefaultDict[
```
Should not remove the if foreach here

ringohoffman replied:
`per_device_and_dtype_grads` is potentially unbound otherwise. Is there a major concern if we define it but don't use it? Is there another alternative so that we can be sure this variable is always bound?

janeyx99 replied:
Would it be considered bound if we explicitly only access per_device_and_dtype_grads when foreach is True? Or would the required change be to rewrite this function in terms of 

if foreach:
    ...blah blah...
else:
   ...blah blah...

If it requires a more involved change to make it bound, I'm okay with defining it and not using it instead of refactoring to much of the code.

ringohoffman replied:
An alternative would be to bind it to `None` `if not foreach` and then assert it's not `None` when you make operations on it:

```python
per_device_and_dtype_grads: Optional[DefaultDict[torch.device, DefaultDict[torch.dtype, List[torch.Tensor]]]
if foreach:
    per_device_and_dtype_grads = defaultdict(lambda: defaultdict(list))
else:
    per_device_and_dtype_grads = None

if foreach:
    assert per_device_and_dtype_grads is not None
    per_device_and_dtype_grads.method(...)
```

it just needs be defined in all paths

janeyx99 replied:
Ah, I like this better.

ringohoffman replied:
Wait, which do you like better? The current implementation?

janeyx99 replied:
Haha I realize I was unclear, defaulting to None--as this is more readable.

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -527,13 +619,13 @@ def add_param_group(self, param_group):
                 raise TypeError("optimizer can only optimize Tensors, "
                                 "but one of the params is " + torch.typename(param))
-            if not self.defaults.get('differentiable', None) and not (param.is_leaf or param.retains_grad):
+            # TODO: https://github.com/pytorch/pytorch/issues/30788
```
this TODO seems irrelevant

ringohoffman replied:
I added this TODO to explain the `type: ignore[attr-defined]` below since `retains_grad ` is not exposed even though it is available and being used. I can remove it if you want.

ringohoffman replied:
I added a clarifying comment about what this is for LMK if you still want me to remove it.

janeyx99 replied:
Instead of the TODO linking to just the issue, could you write a comment instead on why the type ignore is necessary? (e.g., why is it not exposed/where is it used/when we can expect to resolve this). Feel free to still reference the issue but it is not clear why the issue is relevant from having no context.

ringohoffman replied:
Okay so this is what I am going to put instead:

```python
# Tensor.retains_grad is not exposed in the Python API https://github.com/pytorch/pytorch/issues/30788
```

janeyx99 replied:
I'm still failing to see the connection between retains_grad typing and the Jit/C++ issue. Could you elaborate on how they're related?

ringohoffman replied:
I think I misunderstood the purpose of that ticket. My goal was to say that we need to at add `retains_grad` to the definition of `_TensorBase` in the generated  `torch/_C/__init__.pyi`, otherwise static type-checkers have no idea that this attribute exists on `Tensor`s (like we see here).

ringohoffman replied:
I'm not sure when or why it exists or if that is the appropriate place to add it or how it would get added, though.

ringohoffman replied:
I went ahead and removed it but I am curious if you know how we can expose it

janeyx99 replied:
Oh I see, yea the other issue is a really old issue tracking compatibility with JIT...which is now not really maintained. Feel free to add a comment saying "we need to at add retains_grad to the definition of _TensorBase in the generated torch/_C/__init__.pyi, otherwise static type-checkers have no idea that this attribute exists on Tensors" explaining the type ignore

ringohoffman replied:
https://github.com/pytorch/pytorch/issues/103485

janeyx99 replied:
Beautiful

janeyx99 reviewed (commented):
Went over with a first round of comments and questions. Thanks so much for your effort in this!

ringohoffman commented on torch/optim/optimizer.py:
```diff
@@ -167,11 +244,19 @@ class Optimizer:
-        self._optimizer_step_pre_hooks: Dict[int, Callable] = OrderedDict()
-        self._optimizer_step_post_hooks: Dict[int, Callable] = OrderedDict()
+        self._optimizer_step_pre_hooks: Dict[  # type: ignore[valid-type]
+            int, Callable[[Self, Tuple[Any, ...], Dict[str, Any]], Optional[Tuple[Any, Dict[str, Any]]]]
```
Bc you will not get an error in the following example from the description:

```python
optimizer.register_step_post_hook(hook_fn_other_optimizer)  # Parameter 1: type "Adam" cannot be assigned to type "Adagrad"
```

`PostHook` and `PreHook` accept any `Optimizer`. If you define a hook that specifically hints one subclass and then register it on another, you should get an error.

janeyx99 replied:
Oh, in this case, do you think it'd be useful to break up the PreHook and PostHook into HookArgs, HookKwargs, and HookReturnType? Otherwise, the wins here are not very indicative.

ringohoffman replied:
Yes you're right I should definitely try to make aliases for these! Thanks!

ringohoffman commented on torch/optim/optimizer.py:
```diff
@@ -143,7 +216,7 @@ def register_optimizer_step_post_hook(hook: Callable[..., None]) -> RemovableHan
 
     Returns:
-        :class:`torch.utils.hooks.RemoveableHandle`:
+        :class:`torch.utils.hooks.RemovableHandle`:
```
Check out the [Code Spell Checker VS Code Extension](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker)!

ringohoffman commented on torch/optim/optimizer.py:
```diff
@@ -1,30 +1,103 @@
+def _cast(param: torch.Tensor, value: S, key: Hashable = ...) -> S:
+    ...
+
+def _cast(
```
No reason other than it is not a closure, so there isn't an apparent reason to define it nested. IMO it clutters the function it was defined in, especially more so with these overload signatures.

janeyx99 replied:
But we shouldn't need the overload signatures for this, right?

janeyx99 replied:
Even if we move it out of the scope of the previous function, let's keep its definition close to where it's used.

ringohoffman replied:
I'm going to move it to be a `staticmethod` right above `load_state_dict` please let me know what you think of the changes.

ringohoffman commented on torch/utils/_foreach_utils.py:
```diff
@@ -18,22 +18,28 @@
-def _group_tensors_by_device_and_dtype(tensorlistlist: List[List[Tensor]],
-                                       with_indices: Optional[bool] = False) -> \
-        Dict[Tuple[torch.device, torch.dtype], List[List[Union[Tensor, int]]]]:
+def _group_tensors_by_device_and_dtype(
```
beautiful!

ringohoffman replied:
Yes this is so much cleaner!

ringohoffman replied:
Did your change get reverted?

Commit:
Merge branch 'main' into merge-and-improve-torch-optim-optimizer-type-stubs

Commit:
Refactor _group_tensors_by_device_and_dtype hints based on C++ API update

ringohoffman commented on torch/optim/optimizer.py:
```diff
@@ -167,11 +244,19 @@ class Optimizer:
-    def __init__(self, params, defaults):
+    def __init__(
+        self,
+        params: Union[Iterable[torch.Tensor], Iterable[Dict[str, Any]]],
```
This is defined in the type stubs, so I don't think this can be imported here

janeyx99 replied:
Oh hm. Should we just make this into a TypeAlias too and get rid of the type stubs?

ringohoffman replied:
It's not without precedence: https://github.com/search?q=repo%3Apytorch%2Fpytorch+%22_t+%3D+Union%22+path%3A*.py&type=code

Definitely not opposed!

ringohoffman commented on torch/optim/lr_scheduler.pyi:
```diff
@@ -1,5 +1,6 @@
 from typing import Any, Callable, Dict, Iterable, List, Optional, Union
 
+from .common_types import _params_t
```
my bad!

ringohoffman commented on torch/distributed/_spmd/api.py:
```diff
@@ -143,7 +143,7 @@ def _rematerialize_optimizer(
 
     # update opt.state with proxy tensors
-    orig_states: Dict[str, Any] = copy(opt.state)
+    orig_states = copy(opt.state)
```
Because I hinted `state` as `DefaultDict` we get:

```
torch/distributed/_spmd/api.py:160:21: error: Incompatible types in assignment (expression has type "Dict[str, Any]", variable has type "defaultdict[str, Dict[Any, Any]]")  [assignment]
```

But also yes, typically I only really see generic types (like `list`, `dict`, etc.) annotated when they are initialized as empty or the type is more general than what you are initializing it to. Here we can just infer the type from the copy.

janeyx99 replied:
ah. this hints at bigger problems I guess--can we constrain opt.state to be a defaultdict without breaking old code like this?

janeyx99 replied:
Sorry, I had reversed the order in my previous comment. My concern is that if this code (in distributed) made assumptions and added the type for the result of `copy(opt.state)` to be `Dict[str, Any]` and had no problems before but would run into problems now, then there may be user code that do the same thing today and will be broken by this change.

ringohoffman replied:
I understand your concern about breaking things in user code, but I think what we see here is about as much as we can expect user code to "break". This change was purely typing related. Nothing has changed about how this code would run. It was a `defaultdict` before and it still is now, and `defaultdict` is still just a subclass of `dict`.

janeyx99 replied:
Ah, I see state was always a defaultdict, the contention here is not Dict vs DefaultDict, but the value of the dict being Any vs Dict[Any, Any], which doesn't work out with typechecking cuz of contravariance, I suppose?

ringohoffman replied:
Well... pyright doesn't complain that `DefaultDict[str, Dict[Any, Any]]` is being casted down to `Dict[str, Any]`... this seems like a quirk of `mypy` IMO. it is really weirdly strict about this kind of assignments.

Commit:
Remove unused import

https://github.com/pytorch/pytorch/pull/102593#discussion_r1224648448

ringohoffman commented on torch/optim/optimizer.py:
```diff
@@ -1,30 +1,103 @@
+T = TypeVar("T")
+
+@runtime_checkable
+class _ConstructableIterable(Generic[T], Protocol):  # type: ignore[misc]
```
Narrowing something to `collections_abcs.Iterable` : https://docs.python.org/3/library/collections.abc.html#id18

> Checking `isinstance(obj, Iterable)` detects classes that are registered as [Iterable](https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable) or that have an `__iter__()` method,

Defining a [`runtime_checkable`](https://github.com/pytorch/pytorch/blob/803bfdf38a000ce1b189a839925ed192ad75a60b/torch/optim/optimizer.py#L99) `Protcol`:

> `runtime_checkable()` will check only the presence of the required methods or attributes, not their type signatures or types.

allows us to check for both `__iter__` and `__init__` (if it did not either of these, the following line in `_cast` would simply fail).

https://github.com/pytorch/pytorch/blob/803bfdf38a000ce1b189a839925ed192ad75a60b/torch/optim/optimizer.py#L99

ringohoffman replied:
I added it bc once we type this function, we get:

```
torch/optim/optimizer.py:523:24: error: Too many arguments for "Iterable"  [call-arg]
```

ringohoffman replied:
Reading that `isinstance` against `runtime_checkable` `Protocols` can have performance concerns... I think it's fine to just ignore this error and forego this `Protocol`.

janeyx99 replied:
Wait, what would foregoing the protocol look like?

ringohoffman replied:
It's already in! Just added a `type: ignore` to ignore the fact that `Iterable` doesn't have a `__init__`.

Commit:
Create class variable TypeAlias for OptimizerPreHook and OptimizerPostHook

https://github.com/pytorch/pytorch/pull/102593#discussion_r1224713559, https://github.com/pytorch/pytorch/pull/102593#discussion_r1224723159, https://github.com/pytorch/pytorch/pull/102593#discussion_r1224723453

ringohoffman commented on torch/optim/optimizer.py:
```diff
@@ -190,22 +275,22 @@ def __init__(self, params, defaults):
             param_groups = [{'params': param_groups}]
 
         for param_group in param_groups:
+            assert isinstance(param_group, dict)
```
The very first thing that `add_param_group` does is throw an error if `not isinstance(param_group, dict)`, so I don't think this any side effects?

```python
    def add_param_group(self, param_group: Dict[str, Any]) -> None:
        ...
        if not isinstance(param_group, dict):
            raise TypeError(f"param_group must be a dict, but got {type(param_group)}")

```

I am performing the `assert` bc the lines above this:

```python
        if not isinstance(param_groups[0], dict):
            param_groups = [{'params': param_groups}]
```
do not actually convert `List[Tensor]` into `Dict[List[Tensor]]` since type checkers don't support this style of narrowing.

An equivalent assert would be `assert not isinstance(param_groups, torch.Tensor)`, but I was pretty sure this was safe bc of the immediately proceeding `TypeError` on the same condition.

janeyx99 replied:
Would a cast accomplish the same thing here?

ringohoffman replied:
yes

Commit:
Move params_t TypeAlias to optimizer.py

https://github.com/pytorch/pytorch/pull/102593#discussion_r1224788498

Commit:
Move valid-type type: ignore comment

These are fixed in the newest version of mypy

ringohoffman commented on torch/optim/optimizer.py:
```diff
@@ -1,30 +1,103 @@
+    ...
+
+@overload
+def _cast(param: torch.Tensor, value: S, key: Hashable = ...) -> S:
```
Forgive me if I am wrong, but I think you might be misunderstanding `overload`:
* [`@overload`](https://docs.python.org/3.12/library/typing.html#typing.overload)
    * > allows describing functions and methods that support multiple different combinations of argument types
    * intended to handle polymorphic functions whose return type depends on the input
    * I actually got help from Eric Traut, the author of pyright, with typing this specific function: https://github.com/microsoft/pyright/discussions/5199
* [`@final`](https://docs.python.org/3.12/library/typing.html#typing.final)
    * > A decorator to indicate to type checkers that the decorated method cannot be overridden, and the decorated class cannot be subclassed
* [`@override`](https://docs.python.org/3.12/library/typing.html#typing.override)
    * > A decorator for methods that indicates to type checkers that this method should override a method or attribute with the same name on a base class. This helps prevent bugs that may occur when a base class is changed without an equivalent change to a child class

`@overload` here is allowing us to infer different return types based on the type of `value`, not saying that we expect people to "overload"/override this method.

Commit:
Make _cast into a staticmethod

https://github.com/pytorch/pytorch/pull/102593#discussion_r1225034215

Commit:
Add clarifying comment about TODO

Commit:
Remove unnecessary type ignore

Commit:
Remove extra newline

Commit:
Merge branch 'main' into merge-and-improve-torch-optim-optimizer-type-stubs

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -302,22 +338,28 @@ def wrapper(*args, **kwargs):
+        tensorlistlist: List[List[Optional[torch.Tensor]]],
+        with_indices: bool = False,
+    ) -> Union[
+        Dict[Tuple[None, None], Tuple[List[List[Optional[torch.Tensor]]], List[int]]],
```
LOL could we do some TypeAliasing here too, e.g.,

TensorListList would be List[List[Optional[torch.Tensor]]]
Indices would be List[int]

Commit:
Use cast to address BC concerns

Commit:
Add TypeAliases for TensorListList and Indices

https://github.com/pytorch/pytorch/pull/102593#discussion_r1226820859

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -416,44 +503,26 @@ def load_state_dict(self, state_dict):
         # State that is not assigned to params is copied as is (needed for
         # backward compatibility).
-        state = defaultdict(dict)
+        state: DefaultDict[str, Dict[Any, Any]] = defaultdict(dict)
```
```suggestion
        state: DefaultDict[str, Any] = defaultdict(dict)
```
state could be a str to a list of dicts (e.g., with the param_groups key)

This should also not require the deletion of the typing in the _spmd/api.py.

ringohoffman replied:
> This should also not require the deletion of the typing in the _spmd/api.py.

I think the deletion of typing is bc mypy doesn't like casting `DefaultDict` down to `Dict`, so this had no effect:
```
torch/distributed/_spmd/api.py:160:21: error: Incompatible types in assignment (expression has type "Dict[str, Any]", variable has type "defaultdict[str, Any]")  [assignment]
```

ringohoffman replied:
Should we change this to `DefaultDict[str, Union[Dict[Any, Any], List[Dict[Any, Any]]]]`?

janeyx99 replied:
I got bamboozled and thought this was referring to the state_dict for a moment. This is just the state part. I do think the type may be wrong though, see https://github.com/pytorch/pytorch/pull/102593/files#r1227469140.

ringohoffman replied:
What are your thoughts about updating the value types?

janeyx99 replied:
I'm okay with it as Any vs Dict[Any, Any]

janeyx99 commented:
We are almost there!

Commit:
Alternatively bind per_device_and_dtype_grads to None in zero_grad when not foreach

https://github.com/pytorch/pytorch/pull/102593#discussion_r1227410828

Commit:
Remove TODO statement

https://github.com/pytorch/pytorch/pull/102593#discussion_r1224726354

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -188,8 +227,8 @@ def __init__(self, params, defaults):
 
-        self.state = defaultdict(dict)
-        self.param_groups = []
+        self.state: DefaultDict[str, Dict[Any, Any]] = defaultdict(dict)
```
Wait a moment...is the key str here? I believe it is usually a parameter (tensor) when the state gets initialized in _init_group() of the optimizers (see adamw.py for an example).

ringohoffman replied:
oops you're right these definitely look like Tensor keys

janeyx99 replied:
It gets real confusing because by the time it goes to a state_dict,  the representation is usually an int, and when other people superclass this, it can be a fully qualified name (or str).

Commit:
Fix state key type to be Tensor

https://github.com/pytorch/pytorch/pull/102593/files#r1227469140

janeyx99 commented:
By the way, there is some churn in these files recently with some changes getting reverted + others getting landed. e.g., https://github.com/pytorch/pytorch/pull/102619 and the grouping in C++ one. We are hoping to reland the grouping one so don't bother changing that code too fast, but the first PR will affect _cast.

janeyx99 commented:
Alright, the C++ grouping has been relanded! Could you rebase again?

Commit:
Merge branch 'main' into merge-and-improve-torch-optim-optimizer-type-stubs

Commit:
Change state type to DefaultDict[torch.Tensor, Any]

https://github.com/pytorch/pytorch/pull/102593#discussion_r1227469140

Commit:
Add back load_state_dict type hints

ringohoffman commented:
@janeyx99 

I'm taking a look at the type hints for `_process_value_according_to_param_policy`...

`param_groups` is defaulted to `None` but
1. `for pg in param_groups` will fail if `param_groups is None`
2. Neither `cast` nor `_process_value_according_to_param_policy` is ever explicitly called without `param_groups`

`param_id` is defaulted to `None`
1. the entrypoint to `cast` is always called with `param_id=k` where `for k, v in state_dict['state'].items()` and `k in id_map`
2. Neither `cast ` nor `_process_value_according_to_param_policy` is ever explicitly called without `param_id`

The `None` default value doesn't agree with these types and doesn't really seem like it would ever really be true. Would it be okay to remove the `None` default?

janeyx99 commented:
@ringohoffman yes, feel free to remove the defaults.

Commit:
Update _process_value_according_to_param_policy type hints

https://github.com/pytorch/pytorch/pull/102593#issuecomment-1601475990

Commit:
Add back call-arg ignore for Iterable

ringohoffman commented:
lmk what you think of the latest I think we're probably just about there

Commit:
Merge branch 'main' into merge-and-improve-torch-optim-optimizer-type-stubs

ringohoffman commented:
Looks like the [BC linter started failing d.t. the changes to `register_optimizer_step_pre_hook `, `register_optimizer_step_post_hook `, `Optimizer.register_step_pre_hook`, and `Optimizer.register_step_post_hook`](https://github.com/pytorch/pytorch/actions/runs/5480731443/jobs/9984265594#step:2:916), but the types being changed to are more generic than what is currently hinted.

janeyx99 commented:
The BC linter is complaining because it also checks that new code should be usable with old torch versions (and old code should work with new torch versions), and widening or narrowing the number of things that can be considered a Callable would fail in at least 1 direction of these.

Also, going from Callable[Ellipsis, None] to Callable[
    ["Optimizer", Tuple[Any, ...], Dict[str, Any]], Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]
] is narrowing and not widening right? Since before, we wanted any function that takes in whatever but returns None, and now we want a function that takes in 3 things and returns either None or a Tuple. Or maybe I'm not understanding the widening/narrowing correctly.

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -1,29 +1,52 @@
 
+GlobalOptimizerPreHook: TypeAlias = Callable[
+    ["Optimizer", Tuple[Any, ...], Dict[str, Any]], Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]
+]
```
Should we go one step further and have

Args: TypeAlias = Tuple[Any, ...]
Kwargs: TypeAlias = Dict[str, Any]

?

ringohoffman replied:
I think this does improve readability

janeyx99 commented:
About the BC--we can suppress it but call it out as topic: bc-breaking with a label.

Commit:
Add Args and Kwargs TypeAlias

https://github.com/pytorch/pytorch/pull/102593#discussion_r1258470363

ringohoffman commented:
> Also, going from Callable[Ellipsis, None] to Callable[ ["Optimizer", Tuple[Any, ...], Dict[str, Any]], Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]] ] is narrowing and not widening right?

Yes, the parameters have been narrowed. I was thinking about the return type of `Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]` instead of `None`, which was widened.

> About the BC--we can suppress it but call it out as topic: bc-breaking with a label.

I'm not capable of adding labels. Are you able to do this for me?

janeyx99 reviewed (approved):
thank you so much for following through! happy to be migrating away from .pyi and to have nice types in optimizer.py

ringohoffman commented:
@janeyx99 can this failure be retried? doesn't look related as far as I can tell

janeyx99 commented:
I've retried it already, and it doesn't seem to fail on trunk, so I'm not sure if this is irrelevant. https://hud.pytorch.org/hud/pytorch/pytorch/main/1?per_page=50&name_filter=manywheel

@huydhn would you know if the manywheel failure is relevant? It indeed does not look relevant

huydhn commented:
> I've retried it already, and it doesn't seem to fail on trunk, so I'm not sure if this is irrelevant. https://hud.pytorch.org/hud/pytorch/pytorch/main/1?per_page=50&name_filter=manywheel
> 
> @huydhn would you know if the manywheel failure is relevant? It indeed does not look relevant

The failure in https://github.com/pytorch/pytorch/actions/runs/5511497388/jobs/10050396484 is indeed not relevant. I thought that @malfet fixed that on https://github.com/pytorch/builder last week but retrying didn't work is a bit concerning as I see that you have already retried the build job.

malfet commented:
@huydhn I guess this PR just needs a rebase, but considering that everything else is green, merge -i sounds reasonable

malfet commented:
@pytorchbot merge -i

huydhn commented:
Ah, I think I might have some clues here.  Let me delete the artifacts on GitHub and retry just to confirm that it works

malfet commented:
@pytorchbot revert -m "There is nothing wrong with this PR, but it fails some internal builds that depend on outdated typing_extensions, will reland when update is done" -c ghfirst

janeyx99 commented:
@malfet Is this safe to reland?

malfet commented:
@janeyx99 if merge conflicts are resolved it should be (though to err on the side of caution I would've imported the PR and run command that previously failed)

janeyx99 commented:
@ringohoffman Would you like to take charge of resolving the conflicts or shall I take over the PR?

Commit:
Merge branch 'main' into merge-and-improve-torch-optim-optimizer-type-stubs

ringohoffman commented:
@janeyx99 let me know how my merge looks to you!

Commit:
fix mypy errors related to mypy upgrade

janeyx99 commented:
Importing to verify internal things are good

malfet reviewed (approved):
Previous failure no longer happen...

malfet commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge -ic

malfet commented:
@pytorchbot merge -i
